source/kmap_writer.cpp

```diff
@@ -0,0 +1,330 @@
+	auto houseFile = builder.CreateString(getFullFileName(map.getHouseFilename()));
+	auto description = builder.CreateString("Saved with Remere's Map Editor " + __RME_VERSION__ + "\n" + map.getMapDescription());
+
+	return Kmap::CreateMapHeader(
+		builder,
+		map.getWidth(),
+		map.getHeight(),
+		map.getVersion().otbm,
+		monsterSpawnFile,
+		npcSpawnFile,
+		houseFile,
+		description
+	);
+}
+
+FlatOffSetMapData KmapWriter::buildMapData(Map & map)
+{
+	auto areas = buildArea(map);
+	auto towns = buildTowns(map);
+	auto waypoints = buildWaypoints(map);
+
+	return Kmap::CreateMapData(builder, areas, towns, waypoints);
+}
+
+FlatAreaVector KmapWriter::buildArea(Map & map)
+{
+	std::vector<FlatOffSetArea> areasOffset;
+	std::vector<FlatOffSetTile> tilesOffset;
+
+	Position * areaPos = nullptr;
+	MapIterator mapIterator = map.begin();
+	while (mapIterator != map.end())
+	{
+		Tile *tile = (*mapIterator)->get();
+		if (tile == nullptr || tile->size() == 0)
+		{
+			++mapIterator;
+			continue;
+		}
+
+		const Position &pos = tile->getPosition();
+
+		if (isNewArea(pos, areaPos))
+		{
+			std::vector<FlatOffSetTile> finalizedOffset;
+			tilesOffset.swap(finalizedOffset);
+			auto tiles = builder.CreateVector(finalizedOffset);
+
+			auto position = Kmap::CreatePosition(
+				builder,
+				areaPos->x &0xFF00,
+				areaPos->y &0xFF00,
+				areaPos->z
+			);
+
+			areasOffset.push_back(Kmap::CreateArea(builder, tiles, position));
+
+			areaPos->x = pos.x;
+			areaPos->y = pos.y;
+			areaPos->z = pos.z;
+			tilesOffset.clear();
+
+		}
+
+		auto offset = buildTile(map, *tile, pos);
+		if (!offset.IsNull())
+		{
+			tilesOffset.push_back(offset);
+		}
+
+		++mapIterator;
+	}
+
+	return builder.CreateVector(areasOffset);
+}
+
+FlatOffSetTile KmapWriter::buildTile(Map &map, Tile &tile, const Position &pos)
+{
+	Item *ground = tile.ground;
+	if (ground == nullptr || ground->isMetaItem())
+	{
+		return 0;
+	}
+
+	House *house = map.houses.getHouse(tile.getHouseID());
+	auto houseInfoOffset = house ? Kmap::CreateHouseInfo(
+		builder,
+		tile.getHouseID(),
+		house->getEmptyDoorID()
+	) : 0;
+
+	auto createActionOffSet = buildActionAttributes(*ground, pos);
+
+	return Kmap::CreateTile(
+		builder,
+		buildGround(*ground, pos),
+		tile.getX() &0xFF,
+		tile.getY() &0xFF,
+		tile.getMapFlags(),
+		ground->getID(),
+		houseInfoOffset
+	);
+}
+
+bool KmapWriter::isNewArea(const Position &pos, Position *areaPos)
+{
+	if (!areaPos) return false;
+
+	return pos.x < areaPos->x ||
+		pos.x >= areaPos->x + 256 ||
+		pos.y < areaPos->y ||
+		pos.y >= areaPos->y + 256 ||
+		pos.z != areaPos->z;
+}
+
+FlatOffSetItem KmapWriter::buildGround(Item &item, const Position &pos)
+{
+	auto createActionOffSet = buildActionAttributes(item, pos);
+	auto itemAttributesOffset = CreateItemAttributes(
+		builder,
+		0,
+		0,
+		0,
+		0,
+		0,
+		createActionOffSet
+	);
+
+	return Kmap::CreateItem(
+		builder,
+		0,
+		item.getID(),
```

Comment:
Shouldn't be id first then 0?

Reply:
![image](https://user-images.githubusercontent.com/8551443/185603076-9cf9cf15-e8c9-4a92-80e2-40540b3830d0.png)
item id is the second

Reply:
Ok, you're right my bad